libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -630,6 +630,104 @@ string YulUtilFunctions::overflowCheckedIntExpFunction(
 	});
 }
 
+string YulUtilFunctions::overflowCheckedIntLiteralExpFunction(
+	RationalNumberType const& _baseType,
+	IntegerType const& _exponentType,
+	IntegerType const& _commonType
+)
+{
+	solAssert(!_exponentType.isSigned(), "");
+	solAssert(_baseType.isNegative() == _commonType.isSigned(), "");
+	solAssert(_commonType.numBits() == 256, "");
+
+	// Calculates the upperbound for exponentiation, that is, calculate `b`, such that
+	// _base**b <= _maxValue and _base**(b + 1) > _maxValue
+	auto findExponentUpperbound = [](bigint const _base, bigint const _maxValue) -> unsigned
+	{
+		unsigned first = 0;
+		unsigned last = 255;
+		unsigned middle;
+
+		while (first < last)
+		{
+			middle = (first + last)/2;
+
+			if (
+				// The condition on msb is a shortcut that avoids computing large powers in
+				// arbitrary precision.
+				boost::multiprecision::msb(_base) * middle <= boost::multiprecision::msb(_maxValue) &&
```

Comment:
Sorry, I don't really get how this line does what the comment says.

Reply:
The code will work even if we remove `bost::multiprecision::msb(_base) * middle ...` line. However, say `_base = 2**128` and `_maxValue = 2**256 - 1`, in the first iteration of the loop, `middle = 127`. We'll be computing `pow(2**128, 127)` in arbitrary precision, which is a number in `128 * 127 =  16256` bits. But we want to avoid this.

Now, let's define `m = msb(x)` as the largest integer such that `x  >= 2**m`.

If `msb(_base) * middle > msb(_maxValue)`, then `2**(msb(_base) * middle) > 2**msb(_maxValue)`. 

The last statement implies that `2**(msb(_base) * middle) > _maxValue` (from definition of msb) Now `2**(msb(_base) * middle)) = (2**(msb(_base))**middle`. Therefore `_base ** middle >= (2**msb(_base))**middle > _maxValue`.

This means that the condition on power (`pow(_base, middle) <= _maxValue`) is false if the condition on msb is false.

Now, suppose that `msb(_base)*middle <= msb(_maxValue)`, from the definition, `_base < 2**(msb(_base) + 1)` . Therefore, `_base ** middle < 2 ** (msb(_base) * middle + middle) <= 2**(msb(_maxValue) + middle)`. Since `msb(_maxValue) <= 255` and `middle <= 255`, we can say the last value is `<= 2**(255 + 255) <= 2**(510)`.

Therefore, if the condition on the `msb` is true, the computation `pow(_base, middle)` can be done in at most `512` bits. A rough estimate for `pow(_base, middle + 1)` is `512 + 256 = 768` bits, but I think both estimates can be made even smaller.

Reply:
Ok!